addon/adapters/odata.js

```diff
@@ -166,6 +166,76 @@ export default DS.RESTAdapter.extend({
     return Ember.String.pluralize(capitalized);
   },
 
+  /**
+   * A method to make ajax requests.
+   *
+   * @method makeRequest
+   * @param {Object} params
+   * @public
+   */
+  makeRequest(params) {
+    if (!params.method || !params.url) {
+      return;
+    }
+
+    return Ember.$.ajax(params);
+  },
+
+  /**
+   * A method to make any ajax requests.
+   *
+   * @method ajaxAbstraction
+   * @param {Object} params
+   * @param {Function} successCallback
+   * @param {Function} failCallback
+   * @param {Function} alwaysCallback
+   * @return {Promise}
+   * @public
+   */
+  ajaxAbstraction(params, successCallback, failCallback, alwaysCallback) {
+    Ember.assert('Params must be Object!', typeof params === 'object');
+    Ember.assert('params.method or params.url is not defined.', (Ember.none(params.method) && Ember.none(params.url)));
+
+    if (Ember.none(params.method) || Ember.none(params.url)) {
+      throw new Error('params.method or params.url is not defined.');
+    }
+
+    if (params.method === 'POST') {
+      params.data = JSON.stringify(params.data);
+    }
+
+    return Ember.RVSP.Promise(function(resolve, reject) {
+      Ember.$.ajax(params).done((msg) => {
+        if (!Ember.none(successCallback)) {
+          if (typeof successCallback.then == 'function') {
+            successCallback(msg).then(resolve());
+          } else {
+            successCallback(msg);
+            resolve(msg);
+          }
+        }
+      }).fail((msg)=> {
+        if (!Ember.none(failCallback)) {
+          if (typeof failCallback.then == 'function') {
+            failCallback(msg).then(reject(msg));
+          } else {
+            failCallback(msg);
+            reject(msg);
+          }
+        }
+      }).always(()=> {
+        if (!Ember.none(alwaysCallback)) {
+          if (typeof alwaysCallback.then == 'function') {
+            alwaysCallback(msg).then(reject(msg)); //TODO: REJECT OR RESOLVE HERE?
```

Comment:
`alwaysCallback` should be called in `done` and `fail` handlers, because `resolve` or `reject` methods could be called only once. It seems to be right to call `alwaysCallback` after `successCallback` and `failCallback` correspondingly both in sync and async cases.
So `always` handler is unnecessary.